### ISSUE TITLE
Python packages numpy and scipy

### ISSUE THREAD

**Issue by jessdtate** [SCIInstitute/SCIRun#1295]
Users need to be able to easily use numpy and scipy.  It would be best to include them with SCIRun, but we at least need documentation explaining how to do the install/ add to the path.

**Comment by jessdtate** [comment]
installing numy and scipy is very easy (for me) using pip:

cd "scirun_root"/bin/Externals/Install/Python_external/bin/
./python3 -m pip install numpy

similar for scipy, matplotlib and other major packages.

**Comment by aylakhan** [comment]
Which OSs did you test on?

**Comment by jessdtate** [comment]
OS X 10.10.

**Comment by aylakhan** [comment]
Linux will probably work then. This needs testing on Windows.

**Comment by jessdtate** [comment]
There are some issues pushing matlab matrices back to scirun.  They have to be converted to lists of lists, where matlab has an array format.  The conversion is not as straight forward as going the opposite way (at least as far as I have found).  This may be a problem with fields too.

**Comment by dcwhite** [comment]
I should be able to write a converter on the C++ side. What happens when you pass the Matlab array type to the set transient state function?

**Comment by jessdtate** [comment]
nothing.  There is a null output from the module (both the interfacewithpython and pythonobjectforwarder). It doesn't seem to to be waiting around for it, like it finds the object, but can't convert it.

**Comment by dcwhite** [comment]
OK. See if there's (non-python) terminal output--I think I print a message saying the value can't be converted.

**Comment by jessdtate** [comment]
here it is:

[NOTICE] Invalid Connection request: input port is full, or ports are different datatype or same i/o type, or on the same module.

**Comment by dcwhite** [comment]
That's a different message. No worries, I can look at this Monday.

**Comment by jessdtate** [comment]
Oops, That was the one that worked (after conversion).  Here is the error:

No known conversion from python object to C++ object
[PYTHON] 'Transient value set'

**Comment by dcwhite** [comment]
I can't get pip working on my Windows machine. It appears that our python build is lacking required OpenSSL support. Also, we're using the flag `"--with-ensurepip=no"` which seems like it would explain it. Any thoughts?

**Comment by jessdtate** [comment]
I'm not sure, as I'm not an expert in python.  I think that flag would just keep pip out of the default installation, but prevent subsequent installation.  I have had no problems with OS X and pip.

**Comment by aylakhan** [comment]
Turning off ensurepip was due to a Mac OS X build error in 3.4.3. I uploaded 3.4.4 for testing earlier today, since it appears to have some build fixes. OpenSSL on Windows may be tricky.

**Comment by dcwhite** [comment]
@ajanson @jessdtate I got a little bit further on Windows, but we'll probably have to modify the python external build configuration. We decided to push this one until post-IBBM since it will affect so few people.

**Comment by dcwhite** [comment]
I'll focus on getting ensurepip working after upgrading to 3.6 (see #1668).

**Comment by dcwhite** [comment]
Depends on Python upgrade path (gotta get off of 3.4 on Windows first, then ship 3.6 on Mac)

**Comment by dcwhite** [comment]
ensurepip still had problems in 3.8. Not sure why this is so difficult...

**Comment by dcwhite** [comment]
**Investigation findings (2026-06-12)**

The long-standing blocker is that bundled Python is built without OpenSSL, so pip cannot reach PyPI over HTTPS.

**Root cause chain:**
1. `--with-ensurepip=no` in `Superbuild/PythonExternal.cmake` was added to work around a Python 3.4.3 build error (see #1668, now closed). That workaround is no longer needed — `python -m ensurepip` bootstraps pip successfully on the current build.
2. Even with pip bootstrapped, `pip install numpy` fails with `SSL module not available` — all HTTPS fetches are dead because Python was compiled without OpenSSL.

**Current state:**
| Item | Status |
|------|--------|
| Python version (was 3.4, now 3.11.11 default) | ✅ Done |
| `ensurepip` works | ✅ Works (tested locally) |
| OpenSSL linked into bundled Python | ❌ Missing — the blocker |
| numpy/scipy installable | ❌ Blocked on OpenSSL |

**Fix path:**

Two changes to `Superbuild/PythonExternal.cmake`:

1. Change `--with-ensurepip=no` → `--with-ensurepip=install`
2. Add `--with-openssl=<path>` to the UNIX configure flags, detecting OpenSSL per platform:
   - macOS Apple Silicon: `/opt/homebrew/opt/openssl@3`
   - macOS Intel: `/usr/local/opt/openssl@3`
   - Linux: usually found automatically, but can probe with `pkg-config --variable=prefix openssl`
   - Windows: separate handling already needed

3. Optionally, add a `ExternalProject_Add_Step` after the Python build that runs `python -m pip install numpy scipy` so they ship bundled with SCIRun rather than requiring a manual user step.

The ensurepip change is trivial; OpenSSL path detection across platforms is the fiddly part.
